README.md
# flatsheet

> A realtime editor for curating data as a team. [flatsheet.io](http://flatsheet.io)

- **[flatsheet on Github](http://github.com/flatsheet)**
- **[Discussions, tasks, and issue reporting](http://github.com/flatsheet/flatsheet/issues)**
- **[Chat on Gitter](https://gitter.im/flatsheet/flatsheet)**
- **[Contributing guidelines](CONTRIBUTING.md)**

[![Join the chat at https://gitter.im/flatsheet/flatsheet](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/flatsheet/flatsheet?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)

## Current status

This version of Flatsheet is an **in-progress** rewrite using Node.js. The first version of Flatsheet was a [Ruby on Rails prototype](https://github.com/flatsheet/flatsheet-prototype).

Many aspects of the project are incomplete, but things are far enough along that it will be useful if you want to install it and **experiment in a development environment**.

Want to see an early demo of the real-time editor? Go here: http://flatsheet-realtime.herokuapp.com/

### Todos:

Open issues represent current todo list: [github.com/flatsheet/flatsheet/issues](http://github.com/flatsheet/flatsheet/issues)


## Getting started

There are two ways to get flatsheet running:

## Option 1: clone the flatsheet repository

Cloning the repo is best for testing things out & developing flatsheet, but works fine for other purposes. You just have to be prepared to manually pull & merge changes to your server.

### Steps:

- clone this repo: `git clone git@github.com:flatsheet/flatsheet.git`
- change directory: `cd flatsheet`
- run `npm install`
- next you can follow the readme from here: https://github.com/flatsheet/flatsheet#create-an-admin-user

## Option 2: use flatsheet as an npm dependency

This is a good option for when you're using flatsheet in production, because you can make use of npm to install and update flatsheet.

### Steps:

- Create a directory name `flatsheet` for your project (or whatever you like)
- Change directory `cd flatsheet`
- Create a package.json file with `npm init`
- Install flatsheet with npm: `npm install flatsheet --save`
- Install the response module: `npm install response --save`

### File/folder setup
- Create a folder named `data`: `mkdir data`
- Create a app.js file with this code:

```javascript
var server = require('flatsheet')({
  site: {
    title: 'flatsheet',
    email: 'hi@example.com',
    url: 'http://127.0.0.1:3333',
    contact: 'your full name'
  },
  db: __dirname + '/data'
});

server.listen();
```

- Create a .env file for secret config like sendgrid username/password:

```
SENDGRID_USER=yourusername
SENDGRID_PASS=yourpassword
```

- Add a `flatsheet` script and a `start` script to the `scripts` field in your package.json file:

```json
"scripts": {
  "flatsheet": "flatsheet",
  "start": "node app.js"
},
```

- So your full package.json file should look something like this:

```
{
  "name": "flatsheet-example",
  "version": "1.0.0",
  "description": "",
  "main": "app.js",
  "scripts": {
    "flatsheet": "flatsheet",
    "start": "node app.js"
  },
  "author": "",
  "license": "ISC"
}
```

- next create an admin user:


### Create an admin user
- Create an admin account by running `npm run flatsheet accounts create-admin`. You'll be prompted for email, username, & password.
- You can run `npm run flatsheet accounts list` to see that your admin account was created.

### Start the server
- Now run `npm start` to start the server.
- In development you can watch the css & js using `npm run watch`.
- Go to `http://127.0.0.1:3333` and log in with the admin account credentials.

### Create a sheet
- Log in
- Click the **New blank sheet** button
- Fill out the name and description

### Invite users
- Navigate to `http://127.0.0.1:3333/accounts/invite`
- Enter email addresses, one address per line
- Click "Send invitation"
- Users will receive an email with a link they can click to create accounts

## Support

This project is supported in part by a [code sprint grant from OpenNews](http://opennews.org/codesprints.html).

More info [at the Flatsheet blog](http://flatsheet.io/blog/getting-flatsheet-to-v1-with-help-from-opennews/).

## License

[MIT](LICENSE.md)


CHANGELOG.md
# Changelog

## v0.10.0 – April 9, 2015
- API routes for accounts
- Sheets now have owners
- update to latest [table-editor](http://npmjs.org/table-editor) version

## v0.9.2 – April 4, 2015
- fix invites bug

## v0.9.1 – April 3, 2015
- fix `flatsheet` command so it works with accounts refactor

## v0.9.0 – April 3, 2015
- refactor accounts routes into a handlers file and a routes file
- allow custom views so that built-in views can be overriden

## v0.8.0 - March 28, 2015
- sheets now have access control. admins have access to all sheets, users can invite other users to edit their sheets through the sheet settings modal that is added in this version

## v0.7.2 - March 27, 2015
- fix issue with db not being accessible when using flatsheet as a module

## v0.7.1 - March 27, 2015
- limit height of table so horizontal scroll is visible
- reduce whitespace on sheet edit page
- improve html of forms, start using csskit for base css of forms & buttons

## v0.7.0 - March 3, 2015
- revise cli tool: make subcommands plural to they are `accounts` & `sheets`

## v0.6.3 - March 3, 2015
- use 'auth.login' in `/accounts/invite` POST

## v0.6.2 - March 3, 2015
- fix `/accounts/accept` route to be a plural `accounts`

## v0.6.1 - March 3, 2015
- Fix login error: password check
- add CONTRIBUTING.md and LICENSE.md
- small route fixes

## v0.6.0 - February 26, 2015
- make accounts, sheets, sessions routes plural
- update accountdown to ^4.0.0
- allow admins to create, edit, and delete accounts through the ui

## v0.5.2 - February 18, 2015
- Rename util to lib
- move index.js into lib folder
- rename server.js, the example usage of the module, to app.js
- bring back ability to set path to db when using flatsheet as required module

## v0.5.1 - February 18, 2015
- Fix some issues with module names being uppercased, which ubuntu is not fond of.

## v0.5.0 - February 16, 2015
- fixed some install instructions in the readme
- switched from level-sublevel to subleveldown
- switched from level-session to cookie-auth
- added 404 route/view
- use process.env if .env file does not exist
- provide sheet name/description defaults for csv import
- add `flatsheet` to npm scripts so you can run `npm run flatsheet SUBCOMMAND` when you've cloned the repo with git.

## v0.4.0 - December 17, 2014
- allow creating a new sheet by importing a csv

## v0.3.2 - December 17, 2014
- pin level-session to v0.3.1 to avoid incompatibility issues

## v0.3.1 - December 14, 2014
- rename api client package to `flatsheet-api-client`
- fix flatsheet-api-client dependency in `browser/sheet.js`

## v0.3.0 - December 11, 2014
- npm package now named `flatsheet`
- create `server.js` file in directory root so you can get flatsheet running by doing:
  - `git clone`
  - `npm install`
  - `npm start`

## v0.2.8 - November 25, 2014
- add api button to edit view
- update fontawesome to v4.2.0

## v0.2.7 - November 20, 2014
- allow cors requests

## v0.2.6 - November 14, 2014
- now uses minimist for `./bin/flatsheet`
- add the `addView()` method`

## v0.2.2 - November 14, 2014
- now using level-party
- add back public folder

## v0.2.0 - November 14, 2014
- Now requirable & distributed on npm as `flatsheet-server`

## v0.1.0 - November 11, 2014
- Basic functionality works!
- Rewrite from rails/backbone prototype to node/socket.io basically complete!


lib/index.js
var fs = require('fs');
var url = require('url');
var path = require('path');
var http = require('http');
var st = require('st');
var mkdirp = require('mkdirp');
var response = require('response');
var socketio = require('socket.io');
var nodemailer = require('nodemailer');
var sgTransport = require('nodemailer-sendgrid-transport');
var filter = require('filter-object');
var extend = require('extend');
var dotenv = require('dotenv');
var cors = require('corsify');

var Handlebars = require('handlebars');
require('handlebars-layouts')(Handlebars);
require('./handlebars-helpers')(Handlebars);

var initdb = require('./db');

module.exports = Server;

/*
* Main server constructor function
*/

function Server (opts, cb) {
  if (!(this instanceof Server)) return new Server(opts, cb);

  opts = extend({ 
    site: { 
      title: 'flatsheet',
      email: 'hi@flatsheet.io',
      url: 'http://127.0.0.1',
      contact: 'flatsheet admin' 
    }
  }, opts)

  var self = this;
  this.site = opts.site;

  var envFilePath = path.join((opts.dir || process.cwd()), '.env');

  if (fs.existsSync(envFilePath)) {
    var envFile = fs.readFileSync(envFilePath);
    var secrets = dotenv.parse(envFile);
  } else {
    var secrets = {
      SENDGRID_USER: process.env.SENDGRID_USER,
      SENDGRID_PASS: process.env.SENDGRID_PASS
    }
  }

  /*
  * Set path for static files
  */

  this.staticFiles = opts.staticFiles || path.join(__dirname, '..', 'public');
  this.staticFileUrl = opts.staticFileUrl || '/public/'

  mkdirp(this.staticFiles, function (err) {
    if (err) console.error(err)
  });

  /*
  * Set path for db
  */

  this.dataDir = opts.dataDir || opts.db || path.join(__dirname, '..', 'data');

  mkdirp(this.dataDir, function (err) {
    if (err) console.error(err);
    else initdb(self, cb);
  });

  /*
  * Email
  */

  var options = {
    auth: {
      api_user: secrets.SENDGRID_USER,
      api_key: secrets.SENDGRID_PASS
    }
  };

  this.email = nodemailer.createTransport(sgTransport(options));

  /*
  * Set up the application's views
  */

  this.views = {}
  this.viewsDir = path.join(__dirname, '/../views/');
  this.viewData = { site: opts.site };
  this.addViews();

  if (opts.views) {
    if (opts.views[opts.views.length - 1] !== '/')  opts.views += '/';
    this.viewsOverrideDir = opts.views;
    this.overrideViews();
  }

}

/*
*  Get an account based on the session
*/
Server.prototype.getAccountBySession = function (req, cb) {
  var self = this;

  this.auth.getSession(req, function (sessionError, session) {
    if (sessionError) return cb(sessionError);
    
    self.accountdown.get(session.data.username, function (accountError, account) {
      if (accountError) return cb(accountError);
      if (!account) return cb(null, null, session);
      
      var accountFiltered = filter(account, '!email');
      return cb(null, accountFiltered, session);
    });
  });
}


/*
*  listen method for starting the server
*/
Server.prototype.listen = function (port, cb) {
  if (typeof port === 'function') {
    cb = port;
    port = null;
  }

  this.port = port || (process.env.NODE_ENV === 'production' ? 80 : 3333);
  var self = this;

  var cb = cb || function () {
    console.log('flatsheet server listening at ' + self.site.url);
  }

  var staticFiles = st({ path: this.staticFiles, url: this.staticFileUrl });
  var landing = require('../routes/landing')(this);
  var sessions = require('../routes/sessions')(this);
  var accounts = require('../routes/accounts')(this);
  var accountsAPIv20 = require('../routes/accounts-api-v2-0')(this);
  var sheets = require('../routes/sheets')(this);
  var sheetsAPIv20 = require('../routes/sheets-api-v2-0')(this);

  function handler (req, res) {
    if (staticFiles(req, res)) return;
    if (landing.match(req, res)) return;
    if (sessions.match(req, res)) return;
    if (accounts.match(req, res)) return;
    if (accountsAPIv20.match(req, res)) return;
    if (sheets.match(req, res)) return;
    if (sheetsAPIv20.match(req, res)) return;

    self.getAccountBySession(req, function (err, account, session) {
      response().html(self.render('404', { account: account })).pipe(res);
    });
  }

  var opts = {
    'Access-Control-Allow-Origin': '*',
    'Access-Control-Allow-Methods': 'GET, POST, PUT, DELETE',
    'Access-Control-Allow-Headers': 'X-Requested-With, X-HTTP-Method-Override, Content-Type, Accept, Authorization'
  };

  this._server = http.createServer(cors(opts, handler));
  var io = socketio(this._server);
  require('./realtime-editing')(io, this);
  this._server.listen(this.port, cb);
};


/*
*  listen method for stopping the server
*/
Server.prototype.close = function () {
  this._server.close();
  this.db.close();
};


/*
* Create views on application startup
*/
Server.prototype.addViews = function () {
  var self = this;
  
  fs.readdir(this.viewsDir, function (err, files) {
    files.forEach(function (file) {
      if (file === 'layout.html') {
        var filepath = self.viewsDir + 'layout.html';
        Handlebars.registerPartial('layout', fs.readFileSync(filepath, 'utf8'));
      }
      else self.addView(file);
    });
  });
}


/*
* Add a view to the list of views
*/
Server.prototype.addView = function (file, viewsDir) {
  var dir = viewsDir || this.viewsDir;
  return this.views[file.split('.')[0]] = this.compileView(dir + file);
}


/*
* compile a view
*/
Server.prototype.compileView = function (filepath) {
  var template = Handlebars.compile(fs.readFileSync(filepath, 'utf8'));
  return template;
}


/*
* Override views
*/
Server.prototype.overrideViews = function () {
  var self = this;

  fs.readdir(this.viewsOverrideDir, function (err, files) {
    files.forEach(function (file) {
      if (file === 'layout.html') {
        var filepath = self.viewsOverrideDir + 'layout.html';
        Handlebars.registerPartial('layout', fs.readFileSync(filepath, 'utf8'));
      }
      else self.addView(file, self.viewsOverrideDir);
    });
  });
};


/*
* Method for rendering html views
*/
Server.prototype.render = function (view, data) {
  var data = extend(data || {}, this.viewData);
  return this.views[view](data);
};
